Group message data

Niloth-p opened the pull request:
### What does this PR do, and why?

<details>
<summary>Old version</summary>

Refactoring: Group message data in a TypedDict / dataclass.

Group the following into a single piece of data, to simplify passing them as arguments across functions.
- message
- topic_links
- message_links
- time_mentions

Motivation: Was reviewing #1529, the redundancy felt awkward, addressed it with this PR.
Was previously discussed in #1455 comments.

</details>

**Objective**: Avoid passing Message Info data to all of its inner popups.

**Solution Steps:**

1. Introducing a Stacking of Popups
2. Extending PopupView to support a second command to allow both types of exiting.
3. Add a TypedDict to group the content of the message info popup

Why stack popups?
Currently, we create a new popup even when closing the uppermost popups.
With stacking, we can re-use the popups.
This allows us to not pass the entire data for creating the inner popups
to the outer popups.

Support 2 functions.
- Exiting the current popup - goes back to the previous popup.
- Exiting all popups - clears all popups.

Currently, the only popup that opens other popups is the MsgInfoView.
Affected widgets: EditHistoryView, FullRenderedMsgView, FullRawMsgView.

Hotkey behavior (from the widget):
- Press Esc or respective command -> go to MsgInfoView popup
- Press i from the widget -> close all popups (EXIT_POPUP command)

By adding new keypress handling to their parent PopupView, we can
pass their respective commands without having to implement custom
keypress handlers in each of the inner popup classes.


### External discussion & connections
- [ ] Discussed in **#zulip-terminal** in `Stacking of Popups (refactoring Message Info popup) #T1537`
- [ ] Fully fixes #
- [ ] Partially fixes issue #
- [ ] Builds upon previous unmerged work in PR #
- [ ] Is a follow-up to work in PR #
- [ ] Requires merge of PR #
- [x] Merge will enable work on #1455 #1529

### How did you test this?
- [x] Manually - Behavioral changes
- [ ] Manually - Visual changes
- [x] Adapting existing automated tests
- [ ] Adding automated tests for new behavior (or missing tests)
- [x] Existing automated tests should already cover this (*only a refactor of tested code*)

### Self-review checklist for each commit
- [x] It is a [minimal coherent idea](https://github.com/zulip/zulip-terminal#structuring-commits---speeding-up-reviews-merging--development)
- [x] It has a commit summary following the [documented style](https://github.com/zulip/zulip-terminal#structuring-commits---speeding-up-reviews-merging--development) (title & body)
- [x] It has a commit summary describing the  motivation and reasoning for the change
- [x] It individually passes linting and tests
- [ ] It contains test additions for any new behavior
- [x] It flows clearly from a previous branch commit, and/or prepares for the next commit

neiljp reviewed (commented):
@Niloth-p Thanks for looking into this :+1:

This is a quick glance over the code to see if this was the direction I expected; I've not looked at the tests as yet.

It certainly does tidy things up, and makes the code easier to read overall, so appears to be heading in a good direction.

However, one aspect this highlights is how this enables dropping of all but the `self.message` in popups called from the message information popup (eg. topic links). The new `self.msg_data` appears to be used only to get the `self.message` and then bring back the 'parent' message information popup when required.
=> If we didn't have to call `show_msg_info(self.msg_data)`, we could avoid passing all this data around 💡 

We could discuss how to solve the latter issue, since it would be an even cleaner solution, but as above, this tidying would be a good first step in any case :)

@rsashank I'd be interested to hear your feedback on this as it stands, since you have the related PRs ongoing.
- Would this be good merged before or after those?
- Do those use only the `self.message`, or also other data from the message-box in the extra popups?

Niloth-p commented on zulipterminal/ui_tools/views.py:
```diff
@@ -1571,19 +1572,17 @@ class MsgInfoView(PopUpView):
-        topic_links: Dict[str, Tuple[str, int, bool]],
-        message_links: Dict[str, Tuple[str, int, bool]],
-        time_mentions: List[Tuple[str, str]],
+        msg_data: MessageData,
```
Would `MessageInfoPopupContent` and `message_info_content` work better?

Niloth-p committed:
core/views: Stack inner popups, add support for exiting all popups.

Support 2 functions.
- Exiting the current popup - goes back to the previous popup.
- Exiting all popups - clears all popups.

Why stack popups?
Currently, we create a new popup even when closing the uppermost popups.
With stacking, we can re-use the popups.
This allows us to not pass the entire data for creating the inner popups
to the outer popups.

Currently, the only popup that opens other popups is the MsgInfoView.
Affected widgets: EditHistoryView, FullRenderedMsgView, FullRawMsgView.
Updated them to use the newly added functions to exit popups.

Hotkey behavior (from the widget):
- Press Esc or respective command -> go to MsgInfoView popup
- Press i from the widget -> close all popups (EXIT_POPUP command)
By adding new keypress handling to their parent PopupView, we can
pass their respective commands without having to implement custom
keypress handlers.

Tests updated.
Renamed tests:
- test_keypress_show_msg_info -> test_keypress_exit_to_msg_info.
- test_keypress_exit_popup -> test_keypress_exit_all_popups.

Niloth-p committed:
views: Use `self` to re-use instance variables in MsgInfoView class.

And the TestMsgInfoView class.
Tests updated.

Niloth-p committed:
views/core: Delete all message info popup parameters from inner popups.

Now that the inner popups can exit to the outer popup without having to
initialize it, we can remove the parameters of MsgInfoView class from
its inner popup classes and their respective show_* functions.

Tests updated.

Niloth-p committed:
views: Delete the calling arguments from MsgInfoView's show_* calls.

Tests updated.

Niloth-p committed:
refactor: views/helper/core/messages: Group message info in a TypedDict.

Group the following fields that are shown in a message info popup, into
a single piece of data.
- topic_links
- message_links
- time_mentions

Tests updated.
Fixture added.

Niloth-p commented:
Great suggestion, @neiljp! I didn't look into that deeply at the time, thank you for giving me hints to the right direction.
Your awesome explanation helped me come up with this update: Stacking of popups. Hope it's a good solution :crossed_fingers:

rsashank commented:
> @rsashank I'd be interested to hear your feedback on this as it stands, since you have the related PRs ongoing.
> 
> * Would this be good merged before or after those?
> * Do those use only the `self.message`, or also other data from the message-box in the extra popups?

I can rebase that; it shouldn't be a problem if this is merged first. It'll be a bit of work though. :) 

Spoilers use other data as well because we're attempting to return to the message information popup when exiting the SpoilerView. However, the stacking of popups should solve that. I'll look into it.

neiljp commented on zulipterminal/ui_tools/views.py:
```diff
@@ -1585,7 +1585,7 @@ def __init__(
         self.time_mentions = time_mentions
         self.server_url = controller.model.server_url
         date_and_time = controller.model.formatted_local_time(
-            msg["timestamp"], show_seconds=True, show_year=True
+            self.msg["timestamp"], show_seconds=True, show_year=True
```
I see how you've tidied some of the tests here, but I don't understand the intention behind these code changes in this first commit?

neiljp reviewed (commented):
@Niloth-p This is now quite a different PR, with a different tilt towards the refactoring!

I'm sure that (layered) popups can be modeled in many different ways, this was a useful observation that turned out nicely - thanks for changing course and following through :)

The changes look broadly good, and generally are grouped well.

Most of my comments are small points, or thoughts on how it could be improved here and there. We could even hold off on the last commit initially, since I think the earlier ones are a good refactoring all on their own :+1:
(it might also go well when combined with looking at the group of data that comes out of the `soup2markup` end of the processing, which may also benefit from refactoring/simplifying?)

Niloth-p commented on tests/ui_tools/test_popups.py:
```diff
@@ -520,14 +520,14 @@ def test_init(self, msg_box: MessageBox) -> None:
         assert self.full_rendered_message.footer.widget_list == msg_box.footer
 
```
Ah, is that so? I can't seem to recall that discussion.
The change to the commit text sounds better anyways.
Glad to know that the rest is fine.

neiljp replied:
I've not rechecked, but was likely referring to Sashank's comment [here](https://github.com/zulip/zulip-terminal/pull/1537#issuecomment-2248558981), where he is expecting this will be necessary for spoilers.

Niloth-p replied:
Ah, that's covered by this PR. Won't need to pass around any data for that.

Niloth-p commented on tests/ui_tools/test_popups.py:
```diff
@@ -503,19 +503,13 @@ def mock_external_classes(self, mocker: MockerFixture, msg_box: MessageBox) -> N
         self.full_rendered_message = FullRenderedMsgView(
             controller=self.controller,
             message=self.message,
-            topic_links=OrderedDict(),
-            message_links=OrderedDict(),
-            time_mentions=list(),
```
Sorry, thank you!

neiljp commented:
I'm working on refining these points to help with other PRs.
